pallets/interest-accrual/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+				Err(_) => {
+					let new_rate = RateDetails {
+						cumulative_rate: T::InterestRate::saturating_from_rational(100, 100).into(),
+						last_updated: Self::now(),
+					};
+					Rates::<T>::insert(interest_rate_per_sec, &new_rate);
+					new_rate.cumulative_rate
+				}
+				Ok(rate) => {
+					let new_cumulative_rate = Self::calculate_cumulative_rate(
+						interest_rate_per_sec,
+						rate.cumulative_rate,
+						rate.last_updated,
+					)
+					.ok_or(Error::<T>::DebtCalculationFailed)?;
+					// TODO: this should update the rate
+
+					new_cumulative_rate
+				}
+			};
+
+			let debt = Self::calculate_debt(normalized_debt, rate)
+				.ok_or(Error::<T>::DebtCalculationFailed)?;
+			Ok(debt)
+		}
+
+		pub fn do_adjust_normalized_debt(
+			interest_rate_per_sec: T::InterestRate,
+			normalized_debt: T::NormalizedDebt,
+			adjustment: Adjustment<T::Amount>,
+		) -> Result<T::Amount, DispatchError> {
+			let rate =
+				Rates::<T>::try_get(interest_rate_per_sec).map_err(|_| Error::<T>::NoSuchRate)?;
+
+			let debt = Self::calculate_debt(normalized_debt, rate.cumulative_rate)
+				.ok_or(Error::<T>::DebtCalculationFailed)?;
+
+			let new_normalized_debt =
+				Self::convert::<T::InterestRate, T::Amount>(rate.cumulative_rate)
+					.and_then(|rate| {
+						// Apply adjustment to debt
+						match adjustment {
+							Adjustment::Increase(amount) => debt.checked_add(&amount),
+							Adjustment::Decrease(amount) => debt.checked_sub(&amount),
+						}
+						// Calculate normalized debt = debt / cumulative_rate
+						.and_then(|debt| debt.checked_div_int(rate))
+					})
+					.ok_or(Error::<T>::DebtAdjustmentFailed)?;
+
+			Ok(new_normalized_debt)
+		}
+
+		/// Calculates the debt using debt = normalized_debt * cumulative_rate
+		fn calculate_debt(
+			normalized_debt: T::NormalizedDebt,
+			cumulative_rate: T::InterestRate,
+		) -> Option<T::Amount> {
+			// TODO: isn't there a better way of doing this, without the convert?
+			Self::convert::<T::InterestRate, T::NormalizedDebt>(cumulative_rate).and_then(|rate| {
+				normalized_debt
+					.checked_mul(&rate)
+					.and_then(|debt| Self::convert::<T::NormalizedDebt, T::Amount>(debt))
+			})
+		}
+
+		fn calculate_cumulative_rate<Rate: FixedPointNumber>(
+			interest_rate_per_sec: Rate,
+			cumulative_rate: Rate,
+			last_updated: Moment,
+		) -> Option<Rate> {
+			// cumulative_rate * interest_rate_per_sec ^ (now - last_updated)
+			let time_difference_secs = Self::now() - last_updated;
+			checked_pow(interest_rate_per_sec, time_difference_secs as usize)
+				.and_then(|v| v.checked_mul(&cumulative_rate))
+		}
+
+		/// converts a fixed point from A precision to B precision
+		/// we don't convert from un-signed to signed or vice-verse
+		fn convert<A: FixedPointNumber, B: FixedPointNumber>(a: A) -> Option<B> {
+			if A::SIGNED != B::SIGNED {
+				return None;
+			}
+
+			B::checked_from_rational(a.into_inner(), A::accuracy())
+		}
```

Comment:
We also have this in the loans? It might makes sense to move this to commons. Or have dedicated implementation for two generics that are implementing `FixedPointNumber` and use this everywhere. Or will we have a usecase for converting signed-to-un-sigend?

Reply:
This will go away after the Amount -> Balance refactor

Reply:
That would be great :)